### PR TITLE
Improve query field search behavior

### PR DESCRIPTION
**Pull request by cmfcmf**
These are some initial attempts to improve the search query field behavior. I will go into more detail after testing this for a bit.
I have temporarily deployed this branch to dev.adventurelookup.com to play around with it.

<a href="https://gitpod.io/#https://github.com/AdventureLookup/AdventureLookup/pull/308"></a>

### CODE REVIEW

File: app/Resources/webpack/js/adventures_filter/SearchTips.jsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        <h5 className="card-title">Advanced search tips</h5>
+        <p className="card-text">
+          If you enter <strong>magic ghoul party</strong> above, only adventures
+          that contain all three terms will be found. If you want to search for
+          adventures with either <strong>magic</strong>, <strong>ghoul</strong>,
+          or <strong>party</strong>, try searching for{" "}
+          <strong>magic OR ghoul OR party</strong>. Do not use the search bar
+          for things that are available as filters in the sidebar, since using
+          the sidebar filters is more reliable. The search is slightly fuzzy,
+          i.e., searching for <strong>ghoul</strong> will also result in
```
Comment (Wigginns) [inline app/Resources/webpack/js/adventures_filter/SearchTips.jsx:33]: "i.e." should be "e.g." in this case as you are showing an example.

File: app/Resources/webpack/js/adventures_filter/SearchTips.jsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  return (
+    <div className="adl-card">
+      <div className="card-body">
+        <h5 className="card-title">Advanced search tips</h5>
+        <p className="card-text">
+          If you enter <strong>magic ghoul party</strong> above, only adventures
+          that contain all three terms will be found. If you want to search for
+          adventures with either <strong>magic</strong>, <strong>ghoul</strong>,
+          or <strong>party</strong>, try searching for{" "}
+          <strong>magic OR ghoul OR party</strong>. Do not use the search bar
```
Comment (Wigginns) [inline app/Resources/webpack/js/adventures_filter/SearchTips.jsx:30]: Suggest changes "**Do not use** the sidebar for things that are" to "**Avoid using** the search bar for things that are"

**Comment by cmfcmf** [comment]
I am happy with how it works now. I have also added some explanatory comments to the sourc code. Major changes include
- search terms are ANDed by default. To OR them, use ` OR ` in the search query.
- search terms are searched and scored across fields now. Searching for 'ghoul' will now truly return the adventure with the most ghouls in it
- if terms are found in title and description, they are boosted way more than if they are found in, for example, the authors field.

The React changes are performance improvements. Before, the whole sidebar would re-render on every keystroke in the search query box. Caching the callback (`React.useCallback`) and only re-rendering the `Filters` when it is necessary (`React.useMemo`) fix that.

File: src/AppBundle/Service/AdventureSearch.php [hunk 3]
```diff
@@ -397,10 +398,41 @@ private function qMatches(string $q, $matches): array
...
+        // to the same results as searching for 'ghouls galactic'. We could look
+        // into supporting quoting terms ('"galactic ghouls"') later, which would
+        // NOT match adventures with 'ghouls galactic' or adventures with 'galactic'
+        // and 'ghouls' in different fields.
         $clauses = explode(' OR ', $q);
-        $qMatches = [];
+        $orMatches = [];
         foreach ($clauses as $clause) {
             $terms = explode(' ', $clause);
+            // Al terms that are part of this clause have to be ANDed together.
```
Comment (Wigginns) [inline src/AppBundle/Service/AdventureSearch.php:414]:
```suggestion
            // All terms that are part of this clause have to be ANDed together.
```

**Comment by cmfcmf** [comment]
I fixed the typo you noticed, which dismissed your review - it would be great if you could approve this again :+1:

**Pull request merged**